src/runtime/pybuffer.cs

```diff
@@ -0,0 +1,167 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Python.Runtime
+{
+    public class PyBuffer
+    {
+        private IntPtr _handle;
+        private Runtime.Py_buffer _view;
+        private GCHandle _gchandle;
+
+        internal PyBuffer(IntPtr obj, int flags, out bool success)
+        {
+            int size = Marshal.SizeOf(typeof(Runtime.Py_buffer));
+            byte[] rawData = new byte[size];
+            _gchandle = GCHandle.Alloc(rawData, GCHandleType.Pinned);
+            _handle = _gchandle.AddrOfPinnedObject();
+            _view = (Runtime.Py_buffer)Marshal.PtrToStructure(_handle, typeof(Runtime.Py_buffer));
+
+            success = Runtime.PyObject_GetBuffer(obj, _handle, flags) >= 0;
+
+            if (_view.shape != IntPtr.Zero)
+            {
+                Shape = new long[_view.ndim];
+                Marshal.Copy(_view.shape, Shape, 0, (int)_view.len * sizeof(long));
+            }
```

Comment:
When the buffer failed to be retrieved, work on the object continues?

If this failed, IMHO, all operations with side effects from above need to be reversed.

Reply:
If GetBuffer fails, `Shape`, `SubOffsets` and `Strides` are `IntPtr.Zero` anyway.